### PR TITLE
Fix broken and missing Http javadsl methods

### LINKED ISSUE #18017: Akka Http - exposes Tuple2 in javadsl, move to japi.Pair (and java tuples)
Noticed during documenting javadsl of Http:

```
Flow<
  Tuple2<HttpRequest, Integer>,
  Tuple2<Try<HttpResponse>, Integer>,
  HostConnectionPool> poolClientFlow =
  Http.get(system).<Integer>cachedHostConnectionPool("akka.io", 80, materializer);
```

These should be `akka.japi.Pair` (and `akka.japi.tuple.TupleN`) to be consistent with javadsl of streams

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #18017, #18027, #18028.

/cc @sirthias

### CODE REVIEW

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3276/:

```
  9943 Test akka.stream.javadsl.FlowTest.mustBeAbleToRecover failed: java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for 1, took 3.02 sec
  9944     at scala.Predef$.assert(Predef.scala:179)
  9945     at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:338)
  9946     at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:315)
  9947     at akka.testkit.TestKit.expectMsg(TestKit.scala:718)
  9948     at akka.testkit.JavaTestKit.expectMsgEquals(JavaTestKit.java:378)
  9949     at akka.stream.javadsl.FlowTest.mustBeAbleToRecover(FlowTest.java:548)
  9950     ...
 10604 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 10605 Failed tests:
 10606 	akka.stream.javadsl.FlowTest
 13960 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 13961 Total time: 868 s, completed Jul 23, 2015 6:50:22 PM
```

File: akka-docs-dev/rst/java/code/docs/http/javadsl/HttpClientExampleDocTest.java [hunk 1]
```diff
@@ -44,21 +40,20 @@ public void testHostLevelExample() {
...
     final ActorMaterializer materializer = ActorMaterializer.create(system);
 
-    // construct a pool client flow with context type `Int`
-    // TODO these Tuple2 will be changed to akka.japi.Pair
+    // construct a pool client flow with context type `Integer`
     final Flow<
-      Tuple2<HttpRequest, Integer>,
-      Tuple2<Try<HttpResponse>, Integer>,
+      Pair<HttpRequest, Integer>,
+      Pair<Try<HttpResponse>, Integer>,
```
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/HttpClientExampleDocTest.java]: yay

File: akka-docs-dev/rst/java/code/docs/http/javadsl/HttpClientExampleDocTest.java [hunk 2]
```diff
@@ -44,21 +40,20 @@ public void testHostLevelExample() {
...
       HostConnectionPool> poolClientFlow =
       Http.get(system).<Integer>cachedHostConnectionPool("akka.io", 80, materializer);
 
-    // construct a pool client flow with context type `Int`
+    // construct a pool client flow with context type `Integer`
```
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/HttpClientExampleDocTest.java]: good catch

File: akka-docs-dev/rst/java/http/client-side/connection-level.rst [hunk 3]
```diff
@@ -62,7 +62,19 @@ as a more general purpose streaming infrastructure feature.
...
 
 
+.. _http-client-layer-java:
+
 Stand-Alone HTTP Layer Usage
 ----------------------------
 
-// TODO
\ No newline at end of file
+Due to its Reactive-Stream-based nature the Akka HTTP layer is fully detachable from the underlying TCP
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/client-side/connection-level.rst]: nitpick: "Reactive-Streams-based"?

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/WebsocketRoutingExample.java [hunk 4]
```diff
@@ -31,9 +31,9 @@ public Route createRoute() {
      * A handler that treats incoming messages as a name,
      * and responds with a greeting to that name
      */
-    public static Flow<Message, Message, Object> greeter() {
+    public static Flow<Message, Message, ?> greeter() {
```
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/WebsocketRoutingExample.java]: :+1:

File: akka-docs-dev/rst/java/http/client-side/connection-level.rst [hunk 5]
```diff
@@ -62,7 +62,19 @@ as a more general purpose streaming infrastructure feature.
...
+interface. While in most applications this "feature" will not be crucial it can be useful in certain cases to be able
+to "run" the HTTP layer (and, potentially, higher-layers) against data that do not come from the network but rather
+some other source. Potential scenarios where this might be useful include tests, debugging or low-level event-sourcing
+(e.g by replaying network traffic).
+
+On the client-side the stand-alone HTTP layer forms a ``BidiFlow[HttpRequest, SslTlsOutbound, SslTlsInbound, HttpResponse, Unit]``,
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/client-side/connection-level.rst]: `BidiFlow<HttpRequest, SslTlsOutbound, SslTlsInbound, HttpResponse, BoxedUnit>`

File: akka-docs-dev/rst/java/http/client-side/connection-level.rst [hunk 6]
```diff
@@ -62,7 +62,19 @@ as a more general purpose streaming infrastructure feature.
...
+that is a stage that "upgrades" a potentially encrypted raw connection to the HTTP level.
+
+You create an instance of the layer by calling one of the two overloads of the ``Http().clientLayer`` method,
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/client-side/connection-level.rst]: `Http.get(system).clientLayer()`

File: akka-docs-dev/rst/scala/http/low-level-server-side-api.rst [hunk 7]
```diff
@@ -161,7 +161,7 @@ to "run" the HTTP layer (and, potentially, higher-layers) against data that do n
 some other source. Potential scenarios where this might be useful include tests, debugging or low-level event-sourcing
 (e.g by replaying network traffic).
 
-On the server-side the stand-alone HTTP layer forms a ``BidiStage`` that is defined like this:
+On the server-side the stand-alone HTTP layer forms a ``BidiFlow`` that is defined like this:
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/low-level-server-side-api.rst]: good catch!

File: akka-http-core/src/main/scala/akka/http/javadsl/Http.scala [hunk 8]
```diff
@@ -37,6 +40,23 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
...
+    adaptServerLayer(delegate.serverLayer()(materializer))
+
+  /**
+   * Constructs a server layer stage using the given [[ServerSettings]]. The returned [[BidiFlow]] isn't reusable and
+   * can only be materialized once.
+   */
+  def serverLayer(settings: ServerSettings,
+                  remoteAddress: Option[InetSocketAddress],
+                  log: LoggingAdapter,
+                  materializer: Materializer): BidiFlow[HttpResponse, SslTlsOutbound, SslTlsInbound, HttpRequest, Unit] =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: possibly needs more overloads?

File: akka-http-core/src/main/scala/akka/http/javadsl/Http.scala [hunk 9]
```diff
@@ -167,6 +187,20 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
...
+   */
+  def clientLayer(hostHeader: headers.Host): BidiFlow[HttpRequest, SslTlsOutbound, SslTlsInbound, HttpResponse, Unit] =
+    adaptClientLayer(delegate.clientLayer(JavaMapping.toScala(hostHeader)))
+
+  /**
+   * Constructs a client layer stage using the given [[ClientConnectionSettings]].
+   */
+  def clientLayer(hostHeader: headers.Host,
+                  settings: ClientConnectionSettings,
+                  log: LoggingAdapter): BidiFlow[HttpRequest, SslTlsOutbound, SslTlsInbound, HttpResponse, Unit] =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: is 

```
def clientLayer(hostHeader: headers.Host, settings: ClientConnectionSettings)
```

available?

**Comment by ktoso** [comment]
LTGM, just not sure if it's all the overloads we need or if there's one or two missing for japi still

File: akka-http-core/src/main/scala/akka/http/javadsl/Http.scala (hunk 8 above)
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: Yes, I agree. That's only a tiny bit of stupid work more us for the benefit of all those Java users out there.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: Thanks :-)

**Comment by jrudolph** [comment]
Thanks for the comments, @ktoso. I updated the PR using your suggestions.

File: akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala [hunk 10]
```diff
@@ -74,6 +81,13 @@ object JavaMapping {
     }
   }
 
+  /** This trivial mapping isn't enabled by default to prevent it from conflicting with the `Inherited ones `*/
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: last backtick slid away?

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: Yep, thanks, fixed.

**Comment by ktoso** [comment]
LGTM, thanks!

**Pull request merged**